lib/smart_proxy.rb

```diff
@@ -136,24 +128,53 @@ def load_ssl_certificate(path)
       raise e
     end
 
+    def pid_status(pid_path)
+      return :exited unless File.exist?(pid_path)
+      pid = ::File.read(pid_path).to_i
+      return :dead if pid == 0
+      Process.kill(0, pid)
+      :running
+    rescue Errno::ESRCH
+      :dead
+    rescue Errno::EPERM
+      :not_owned
+    end
+
+    def check_pid
+      case pid_status(pid_path)
+      when :running, :not_owned
+        logger.error "A server is already running. Check #{pid_path}"
+        exit(1)
+      when :dead
+        File.delete(pid_path)
+      end
+    end
+
+    def write_pid
+      FileUtils.mkdir_p(File.dirname(pid_path)) unless File.exist?(pid_path)
+      File.open(pid_path, ::File::CREAT | ::File::EXCL | ::File::WRONLY){|f| f.write("#{Process.pid}") }
+      at_exit { File.delete(pid_path) if File.exist?(pid_path) }
+    rescue Errno::EEXIST
+      check_pid
+      retry
+    end
+
     def launch
       ::Proxy::Plugins.configure_loaded_plugins
 
-      create_pid_dir
       http_app = http_app()
       https_app = https_app()
       raise Exception.new("Both http and https are disabled, unable to start.") if http_app.nil? && https_app.nil?
 
-      Process.daemon if SETTINGS.daemon
+      if SETTINGS.daemon
+        check_pid
+        Process.daemon
+        write_pid
+      end
 
       t1 = Thread.new { https_app.start } unless https_app.nil?
       t2 = Thread.new { http_app.start } unless http_app.nil?
 
-      sleep 5 # Rack installs its own trap; Sleeping for 5 secs insures we overwrite it with our own
-      trap(:INT) do
-        exit(0)
-      end
```

Comment:
The delay is needed to ensure that rack finished its initialization before we install our own interrupt handler.